internal/supervisor/installation.go

```diff
@@ -0,0 +1,340 @@
+	clusterInstallations, err := s.store.GetClusterInstallations(&model.ClusterInstallationFilter{
+		PerPage:        model.AllPerPage,
+		InstallationID: installation.ID,
+		IncludeDeleted: true,
+	})
+	if err != nil {
+		logger.WithError(err).Warn("Failed to find cluster installations")
+		return installation.State
+	}
+
+	clusterInstallationIDs := []string{}
+	if len(clusterInstallations) > 0 {
+		for _, clusterInstallation := range clusterInstallations {
+			clusterInstallationIDs = append(clusterInstallationIDs, clusterInstallation.ID)
+		}
+
+		clusterInstallationLocks := newClusterInstallationLocks(clusterInstallationIDs, instanceID, s.store, logger)
+		if !clusterInstallationLocks.TryLock() {
+			logger.Debugf("Failed to lock %d cluster installations", len(clusterInstallations))
+			return installation.State
+		}
+		defer clusterInstallationLocks.Unlock()
+
+		// Fetch the same cluster installations again, now that we have the locks.
+		clusterInstallations, err = s.store.GetClusterInstallations(&model.ClusterInstallationFilter{
+			PerPage:        model.AllPerPage,
+			IDs:            clusterInstallationIDs,
+			IncludeDeleted: true,
+		})
+		if err != nil {
+			logger.WithError(err).Warnf("Failed to fetch %d cluster installations by ids", len(clusterInstallations))
+			return installation.State
+		}
+
+		if len(clusterInstallations) != len(clusterInstallationIDs) {
+			logger.Warnf("Found only %d cluster installations after locking, expected %d", len(clusterInstallations), len(clusterInstallationIDs))
+		}
+	}
+
+	deletingClusterInstallations := 0
+	deletedClusterInstallations := 0
+	failedClusterInstallations := 0
+
+	for _, clusterInstallation := range clusterInstallations {
+		switch clusterInstallation.State {
+		case model.ClusterInstallationStateCreationRequested:
+		case model.ClusterInstallationStateCreationFailed:
+
+		case model.ClusterInstallationStateDeletionRequested:
+			deletingClusterInstallations++
+			continue
+
+		case model.ClusterInstallationStateDeletionFailed:
+			// Only count failed cluster installations if the deletion is in
+			// progress.
+			if installation.State == model.InstallationStateDeletionInProgress {
+				failedClusterInstallations++
+				continue
+			}
+
+			// Otherwise, we try the deletion again below.
+
+		case model.ClusterInstallationStateDeleted:
+			deletedClusterInstallations++
+			continue
+
+		case model.ClusterInstallationStateStable:
+
+		default:
+			logger.Errorf("Cannot delete installation with cluster installation in state %s", clusterInstallation.State)
+			return model.InstallationStateDeletionFailed
+		}
+
+		clusterInstallation.State = model.ClusterInstallationStateDeletionRequested
+		err = s.store.UpdateClusterInstallation(clusterInstallation)
+		if err != nil {
+			logger.WithError(err).Warnf("Failed to mark cluster installation %s for deletion", clusterInstallation.ID)
+			return installation.State
+		}
+
+		deletingClusterInstallations++
+	}
+
+	logger.Debugf(
```

Comment:
Good place for `WithFields()` in my opinion.

Reply:
2/5, but I'm not sure fields are fungible for sprintf-statements in all cases. Certainly we'd want structured logging for pretty much all common verbiage (e.g. the `cluster` and `installation` we do today), but are we actually looking to "parse" `cluster_installation_deleted_count`, etc.?

Reply:
In my head the fields would be no different that what you had: `deleting`, `failed`, etc. You would narrow the search down with a second criteria something else in the log message if you just wanted these messages.

We can always implement this later if we feel like it!